### DISCUSSION TITLE
Server Side Lazy Loading

### DISCUSSION THREAD

**Discussion by alfl** [remix-run/remix#5226] [Q&A]
Love Remix!

We're using a runtime environment that has tight startup time requirements (Cloudflare Workers -- 200ms startup from invocation to event handler registration). Here's our code: https://github.com/kubelt/kubelt/blob/main/apps/passport/server.js

Line 2: `import * as build from "@remix-run/dev/server-build";` often (~95% of runs) takes long enough that we blow the startup time limit. When running locally we're able to do this import asynchronously and can boot 100% of the time, unfortunately Cloudflare does not support async imports in the worker runtime (yet).

Within Remix's features / config / etc. is there a way to make this run faster (bake it to be completely static initially), or to delay loading until after event handler registration?

**Comment by alfl** [comment]
What is done on import, for example, and how can we make it faster / delay it?

**Comment by sergiodxa** [comment]
If the runtime doesn't support dynamic imports I don't think there's anything Remix could do to add it, I think this is something Cloudflare should support.

What if you run a second process to bundle the request handler you export and the Remix build code? So you have a single JS file with everything.

**Reply by re-thc** [reply]
Remix can at least speed this up or trim down what's required here.

**Reply by sergiodxa** [reply]
How could a framework speed up the time the runtime takes to import a module?

And the only way to trim down the server build code would be to minify it, Remix already do it, so if the code is too big it's most likely app code.

I think the solution is to bundle the Cloudflare Worker function code and the Remix's server build code into a single file so you don't need to do any import, bundle everything together.

Maybe Remix could do this, but right now it should be possible to run a second process to do this.

**Reply by re-thc** [reply]
> How could a framework speed up the time the runtime takes to import a module?

By reducing the size of the module and the code required?

**Reply by sergiodxa** [reply]
There’s a limit of how much Remix could reduce its code without removing features.

I also deploy to Cloudflare and I haven’t seen this error so this is most likely because the app code itself is large, if the build code is already minified and it’s still so big that the runtime times out before completing the import then then the only way to solve it is to either avoid the import (by bundling it) or reduce the app code itself.

**Comment by RossMcMillan92** [comment]
Seeing this too. I have a UI-heavy webapp that was recently around 1MB in size. This deployed to CF Pages around 80% of the time. I've recently made changes and that bundle jumped to 2MB, and now it won't deploy at all. Was hoping that using React.lazy and suspense would bring it down, but it bumped the bundle up to 2.6MB. There's a bunch of code that could be client-only but I've no idea how to get it out the server bundle. 

Any advice would be helpful, and I'll report back if I find anything. Thinking I'll contact CF and ask them to adjust my limits as a last resort, as [they seem open to that](https://docs.google.com/forms/d/e/1FAIpQLSd_fwAVOboH9SlutMonzbhCxuuuOmiU1L_I5O2CFbXf_XXMRg/viewform).

<img width="700" alt="A screenshot showing a list of failed CF pages deploys" src="https://user-images.githubusercontent.com/3329319/214681060-fefecdca-8e4c-4fb2-9f44-e9869be1ece1.png">

**Comment by jfsiii** [comment]
We're running into this issue at work due to some actions that use libs for CSV, PDF, and XLS files.

The error from Cloudflare says the issue is from doing too much work at startup and advises moving the work into the event handler. We could do this with a dynamic import in the action, but Cloudflare only supports that in the Module Workers format, and Remix doesn't doesn't support that https://github.com/remix-run/remix/issues/764

The only other option I can think of for now is moving those heavy routes to new workers. It'd work, but I'm not sure how big a pain it'd be to develop & deploy as the number increases.

Along those lines, I wonder if Remix outputting multiple handlers vs one huge one would help? Right now all the code for the entire (server side part of the) app is loaded for any route.

**Comment by re-thc** [comment]
https://blog.cloudflare.com/making-cloudflare-for-web/ cloudflare is increasing startup time limits soon. Maybe it'll help :)
